File: bindings/java/src/org/causalontology/Jcs.java
package org.causalontology;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * RFC 8785 (JSON Canonicalization Scheme) serialization over the object
 * graph produced by Json.parse (LinkedHashMap / ArrayList / String /
 * Boolean / Long / Double / null).
 *
 * Object keys are sorted by UTF-16 code units, which is exactly what
 * String.compareTo does. Strings use the minimal-escape rules. Numbers
 * mirror the Python binding's _jcs_number: a Long prints as its decimal
 * string; a Double that is integer-valued with magnitude below 1e21 prints
 * as an integer; other Doubles print via the shortest round-trip decimal
 * (Double.toString) with the exponent normalized to the ECMAScript form
 * (1.0E-7 becomes 1e-7; 1.0E21 becomes 1e+21). As with the Python binding,
 * full ECMAScript formatting for extreme magnitudes outside the range the
 * conformance suite exercises is pinned at the 1.0.0 freeze.
 */
public final class Jcs {

    private Jcs() {
    }

    /** The canonical serialization of a value. */
    public static String serialize(Object value) {
        StringBuilder sb = new StringBuilder();
        append(value, sb);
        return sb.toString();
    }

    private static void append(Object value, StringBuilder sb) {
        if (value == null) {
            sb.append("null");
            return;
        }
        if (value instanceof Boolean) {
            sb.append(((Boolean) value).booleanValue() ? "true" : "false");
            return;
        }
        if (value instanceof String) {
            sb.append(quote((String) value));
            return;
        }
        if (value instanceof Number) {
            sb.append(number((Number) value));
            return;
        }
        if (value instanceof List) {
            sb.append('[');
            List<?> list = (List<?>) value;
            for (int i = 0; i < list.size(); i++) {
                if (i > 0) {
                    sb.append(',');
                }
                append(list.get(i), sb);
            }
            sb.append(']');
            return;
        }
        if (value instanceof Map) {
            Map<?, ?> map = (Map<?, ?>) value;
            List<String> keys = new ArrayList<>();
            for (Object key : map.keySet()) {
                if (!(key instanceof String)) {
                    throw new IllegalArgumentException(
                        "cannot canonicalize a non-string key: " + key);
                }
                keys.add((String) key);
            }
            // String.compareTo is comparison by UTF-16 code units,
            // which is the RFC 8785 key ordering.
            Collections.sort(keys);
            sb.append('{');
            boolean first = true;
            for (String key : keys) {
                if (!first) {
                    sb.append(',');
                }
                first = false;
                sb.append(quote(key));
                sb.append(':');
                append(map.get(key), sb);
            }
            sb.append('}');
            return;
        }
        throw new IllegalArgumentException(
            "cannot canonicalize value of type " + value.getClass().getName());
    }

    /** The RFC 8785 minimal-escape string form, including the quotes. */
    static String quote(String s) {
        StringBuilder sb = new StringBuilder();
        sb.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':
                    sb.append("\\\"");
                    break;
                case '\\':
                    sb.append("\\\\");
                    break;
                case '\b':
                    sb.append("\\b");
                    break;
                case '\t':
                    sb.append("\\t");
                    break;
                case '\n':
                    sb.append("\\n");
                    break;
                case '\f':
                    sb.append("\\f");
                    break;
                case '\r':
                    sb.append("\\r");
                    break;
                default:
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
                    break;
            }
        }
        sb.append('"');
        return sb.toString();
    }

    /** The RFC 8785 number form, mirroring the Python _jcs_number. */
    static String number(Number n) {
        // Integral types keep their exact decimal representation.
        if (!(n instanceof Double) && !(n instanceof Float)) {
            return n.toString();
        }
        double d = n.doubleValue();
        if (Double.isNaN(d) || Double.isInfinite(d)) {
            throw new IllegalArgumentException(
                "NaN and Infinity are not permitted (RFC 8785)");
        }
        if (d == 0.0) {
            // Covers -0.0 as well: RFC 8785 serializes it as "0".
            return "0";
        }
        if (d == Math.floor(d) && Math.abs(d) < 1e21) {
            // Integer-valued double: print as an integer. BigDecimal is
            // exact for integral doubles and avoids long overflow near 1e21.
            return new BigDecimal(d).toBigInteger().toString();
        }
        // Shortest round-trip decimal.
        String r = Double.toString(d);
        int e = r.indexOf('E');
        if (e < 0) {
            return r;
        }
        // Normalize Java's exponent form to the ECMAScript style:
        // "1.0E-7" -> "1e-7", "1.0E21" -> "1e+21", "1.25E22" -> "1.25e+22".
        String mantissa = r.substring(0, e);
        String exponent = r.substring(e + 1);
        if (mantissa.endsWith(".0")) {
            mantissa = mantissa.substring(0, mantissa.length() - 2);
        }
        String sign = "+";
        if (exponent.startsWith("-")) {
            sign = "-";
            exponent = exponent.substring(1);
        } else if (exponent.startsWith("+")) {
            exponent = exponent.substring(1);
        }
        // Strip any leading zero padding from the exponent digits.
        int firstNonZero = 0;
        while (firstNonZero < exponent.length() - 1
                && exponent.charAt(firstNonZero) == '0') {
            firstNonZero++;
        }
        exponent = exponent.substring(firstNonZero);
        return mantissa + "e" + sign + exponent;
    }
}

File: bindings/go/v4/causalontology/jcs.go
// RFC 8785 (JSON Canonicalization Scheme) serialization.
//
// Object keys are sorted with sort.Strings; for the ASCII keys the
// Causalontology schemas allow, byte order and the UTF-16 code-unit order
// RFC 8785 prescribes coincide. Strings use the minimal escape set. Numbers
// follow the ECMAScript Number-to-string rules: an integer source literal
// is emitted verbatim, and every floating value is rendered from its
// shortest round-trip decimal: printed as a plain integer when it is
// integral with magnitude below 1e21, in plain decimal notation down to
// 1e-4, and in normalized exponent form otherwise (e-7, not e-07; e+21
// keeps its plus sign). This mirrors _jcs_number in the Python binding's
// canonical.py; full ECMAScript exponent formatting for extreme
// magnitudes is pinned at the 1.0.0 conformance freeze, as there.
package causalontology

import (
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"math/big"
	"sort"
	"strconv"
	"strings"
)

// SerializeJCS renders one JSON value in its RFC 8785 canonical form.
func SerializeJCS(value any) (string, error) {
	var builder strings.Builder
	if err := appendJCS(&builder, value); err != nil {
		return "", err
	}
	return builder.String(), nil
}

// JSONEqual reports semantic equality of two JSON values by comparing
// their canonical serializations, so 1 and 1.0 compare equal exactly as
// they do under Python's == on parsed JSON.
func JSONEqual(a, b any) bool {
	canonicalA, errA := SerializeJCS(a)
	canonicalB, errB := SerializeJCS(b)
	if errA != nil || errB != nil {
		return false
	}
	return canonicalA == canonicalB
}

// appendJCS writes the canonical form of one value onto the builder.
func appendJCS(builder *strings.Builder, value any) error {
	switch v := value.(type) {
	case nil:
		builder.WriteString("null")
	case bool:
		if v {
			builder.WriteString("true")
		} else {
			builder.WriteString("false")
		}
	case string:
		builder.WriteString(jcsQuote(v))
	case json.Number, int, int64, float64:
		rendered, err := jcsNumber(v)
		if err != nil {
			return err
		}
		builder.WriteString(rendered)
	case []any:
		builder.WriteByte('[')
		for i, item := range v {
			if i > 0 {
				builder.WriteByte(',')
			}
			if err := appendJCS(builder, item); err != nil {
				return err
			}
		}
		builder.WriteByte(']')
	case map[string]any:
		keys := make([]string, 0, len(v))
		for key := range v {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		builder.WriteByte('{')
		for i, key := range keys {
			if i > 0 {
				builder.WriteByte(',')
			}
			builder.WriteString(jcsQuote(key))
			builder.WriteByte(':')
			if err := appendJCS(builder, v[key]); err != nil {
				return err
			}
		}
		builder.WriteByte('}')
	default:
		return fmt.Errorf("cannot canonicalize a value of type %T", value)
	}
	return nil
}

// jcsQuote renders a string with the RFC 8785 minimal escapes: the seven
// short escapes, \u00xx (lowercase) for the remaining control characters,
// and everything else verbatim.
func jcsQuote(text string) string {
	var builder strings.Builder
	builder.WriteByte('"')
	for _, r := range text {
		switch r {
		case '"':
			builder.WriteString(`\"`)
		case '\\':
			builder.WriteString(`\\`)
		case '\b':
			builder.WriteString(`\b`)
		case '\t':
			builder.WriteString(`\t`)
		case '\n':
			builder.WriteString(`\n`)
		case '\f':
			builder.WriteString(`\f`)
		case '\r':
			builder.WriteString(`\r`)
		default:
			if r < 0x20 {
				fmt.Fprintf(&builder, `\u%04x`, r)
			} else {
				builder.WriteRune(r)
			}
		}
	}
	builder.WriteByte('"')
	return builder.String()
}

// jcsNumber renders any numeric value form canonically. An integer-source
// json.Number is emitted verbatim ("6" stays "6"); a decimal-source
// json.Number ("6.000") and every native float go through the ES6 float
// rules; native ints print in plain decimal.
func jcsNumber(value any) (string, error) {
	switch n := value.(type) {
	case json.Number:
		if IsIntegerNumber(n) {
			return n.String(), nil
		}
		f, err := n.Float64()
		if err != nil {
			return "", err
		}
		return jcsFloat(f)
	case int:
		return strconv.Itoa(n), nil
	case int64:
		return strconv.FormatInt(n, 10), nil
	case float64:
		return jcsFloat(n)
	}
	return "", fmt.Errorf("cannot canonicalize a number of type %T", value)
}

// jcsFloat renders a float64 from its shortest round-trip digits, exactly
// as the Python binding's _jcs_number does: an integral value below 1e21
// prints as a plain integer; decimal notation covers magnitudes down to
// 1e-4 (Python repr's threshold - every float at or above 1e16 is
// integral, so the two implementations agree everywhere); anything
// outside that range uses the normalized exponent form (e-7, not e-07;
// e+21 keeps its plus sign).
func jcsFloat(f float64) (string, error) {
	if math.IsNaN(f) || math.IsInf(f, 0) {
		return "", errors.New("NaN and Infinity are not permitted (RFC 8785)")
	}
	if f == 0 {
		// Both zeros serialize as "0" (ES6 String(-0) is "0").
		return "0", nil
	}
	if f == math.Trunc(f) && math.Abs(f) < 1e21 {
		// An integral value below 1e21 prints as its exact integer,
		// Python's str(int(n)); big.Float carries a float64 exactly, so
		// this stays exact beyond the int64 range too.
		integer, _ := big.NewFloat(f).Int(nil)
		return integer.String(), nil
	}
	sign := ""
	if f < 0 {
		sign = "-"
		f = -f
	}
	// The shortest round-trip decimal in exponent form: "d[.ddd]e±XX".
	shortest := strconv.FormatFloat(f, 'e', -1, 64)
	markerIndex := strings.IndexByte(shortest, 'e')
	mantissa := shortest[:markerIndex]
	exponent, err := strconv.Atoi(shortest[markerIndex+1:])
	if err != nil {
		return "", err
	}
	digits := strings.Replace(mantissa, ".", "", 1)
	// k is the number of significant digits; n is the position of the
	// decimal point relative to the start of the digits (ES6 7.1.12.1).
	k := len(digits)
	n := exponent + 1
	switch {
	case k <= n && n <= 21:
		// Defensive only: integral values below 1e21 were already printed
		// exactly above, so no reachable value lands here.
		return sign + digits + strings.Repeat("0", n-k), nil
	case 0 < n && n <= 21:
		// A decimal point inside the digits.
		return sign + digits[:n] + "." + digits[n:], nil
	case -3 <= n && n <= 0:
		// A small decimal: leading "0." and padding zeros, down to 1e-4
		// (the same threshold Python's repr uses).
		return sign + "0." + strings.Repeat("0", -n) + digits, nil
	default:
		// Exponent form with the normalized exponent: e-7, e+21.
		rendered := digits[:1]
		if k > 1 {
			rendered += "." + digits[1:]
		}
		exponentValue := n - 1
		exponentSign := "+"
		if exponentValue < 0 {
			exponentSign = "-"
			exponentValue = -exponentValue
		}
		return sign + rendered + "e" + exponentSign + strconv.Itoa(exponentValue), nil
	}
}
